### PR TITLE
New account routings system

### PR DESCRIPTION
**Pull request by GuillaumeKergreis**
Replace the old account routing system which only allowed one custom account routing per account (+ deprecated IBAN field).
The new account routing system ass an account routing table to the database to set as many account routing as you want.

### CODE REVIEW

File: obp-api/src/main/scala/code/api/berlin/group/v1_3/JSONFactory_BERLIN_GROUP_1_3.scala [hunk 1]
```diff
@@ -343,7 +344,7 @@ object JSONFactory_BERLIN_GROUP_1_3 extends CustomJsonFormats {
   }
 
   private def getIbanAndBban(x: BankAccount) = {
-    val iBan = if (x.accountRoutings.headOption.isDefined && x.accountRoutings.head.scheme == "IBAN") x.accountRoutings.head.address else x.iban.getOrElse("")
+    val iBan = if (x.accountRoutings.exists(_.scheme == AccountRoutingScheme.IBAN.toString)) x.accountRoutings.find(_.scheme == AccountRoutingScheme.IBAN.toString).map(_.address).getOrElse("") else ""
```
[the remark this answers is not in the document]
Comment (GuillaumeKergreis) [reply obp-api/src/main/scala/code/api/berlin/group/v1_3/JSONFactory_BERLIN_GROUP_1_3.scala]: I changed this in my amended commit.

File: obp-api/src/main/scala/code/bankconnectors/akka/actor/SouthSideActorOfAkkaConnector.scala [hunk 2]
```diff
@@ -106,15 +106,12 @@ object Transformer {
...
-      iban = None,
       number = acc.number,
       bankId = acc.bankId,
       lastUpdate = acc.lastUpdate,
       branchId = acc.branchId,
-      accountRoutingScheme = acc.accountRoutingScheme,
-      accountRoutingAddress = acc.accountRoutingAddress,
-      accountRoutings = Nil,
-      accountRules = Nil,
+      accountRoutings = acc.accountRoutings,
```
[the remark this answers is not in the document]
Comment (GuillaumeKergreis) [reply obp-api/src/main/scala/code/bankconnectors/akka/actor/SouthSideActorOfAkkaConnector.scala]: I've done this :
`def bankAccount(acc: BankAccount): BankAccountCommons = acc`

File: obp-api/src/main/scala/code/bankconnectors/akka/actor/SouthSideActorOfAkkaConnector.scala [hunk 3]
```diff
@@ -165,13 +162,10 @@ object Transformer {
...
         name = t.thisAccount.name,
         label = t.thisAccount.label,
-        iban = t.thisAccount.iban,
         number = t.thisAccount.number,
         bankId = t.thisAccount.bankId,
         lastUpdate = t.thisAccount.lastUpdate,
         branchId = t.thisAccount.branchId,
-        accountRoutingScheme = t.thisAccount.accountRoutingScheme,
-        accountRoutingAddress = t.thisAccount.accountRoutingAddress,
         accountRoutings = t.thisAccount.accountRoutings,
```
[the remark this answers is not in the document]
Comment (GuillaumeKergreis) [reply obp-api/src/main/scala/code/bankconnectors/akka/actor/SouthSideActorOfAkkaConnector.scala]: It's actually better. Done.

**Comment by GuillaumeKergreis** [comment]
I corrected / added some tests for this feature, so I'll need additional review on the three last commits 

- feature/Add a check on existing account routing
- test/Add integration tests for create/update account on account routings
- test/Correct and add some tests for new account routings

All the tests are now passing on Jenkins :
[Jenkins Tests results](https://jenkins.tesobe.com/job/Build-OBP-API-guillaume-jdk13_UNLICENSED-JDK-NOT-FOR-DEPLOYMENT/10/)

File: obp-api/src/main/scala/code/bankconnectors/LocalMappedConnector.scala [hunk 4]
```diff
@@ -1316,19 +1293,50 @@ object LocalMappedConnector extends Connector with MdcLoggable {
...
+    val oldAccountRoutings = BankAccountRouting.findAll(By(BankAccountRouting.BankId, bankId.value),
+      By(BankAccountRouting.AccountId, accountId.value))
+      .map(_.accountRouting)
+
+    // Add or update new routing schemes
+    accountRoutings.foreach(accountRouting =>
+      oldAccountRoutings.find(_.scheme == accountRouting.scheme) match {
+        case Some(_) =>
+          BankAccountRouting
+            .find(By(BankAccountRouting.BankId, bankId.value), By(BankAccountRouting.AccountId, accountId.value),
```
Comment (oldbig) [inline obp-api/src/main/scala/code/bankconnectors/LocalMappedConnector.scala:1309]: duplicated query.
change demo at blow comment.

File: obp-api/src/main/scala/code/bankconnectors/LocalMappedConnector.scala [hunk 5]
```diff
@@ -1316,19 +1293,50 @@ object LocalMappedConnector extends Connector with MdcLoggable {
...
+            .saveMe()
+      }
+    )
+
+    // Delete non-present routing schemes
+    oldAccountRoutings.foreach(accountRouting =>
+      accountRoutings.find(_.scheme == accountRouting.scheme)
+        .getOrElse(
+          BankAccountRouting
+            .find(By(BankAccountRouting.BankId, bankId.value), By(BankAccountRouting.AccountId, accountId.value),
```
Comment (oldbig) [inline obp-api/src/main/scala/code/bankconnectors/LocalMappedConnector.scala:1327]: duplicated query, change demo at blow comment.

File: obp-api/src/main/scala/code/bankconnectors/LocalMappedConnector.scala [hunk 6]
```diff
@@ -1316,19 +1293,50 @@ object LocalMappedConnector extends Connector with MdcLoggable {
...
+              By(BankAccountRouting.AccountRoutingScheme, accountRouting.scheme))
+            .map(_.delete_!)
+        )
+    )
+
```
Comment (oldbig) [inline obp-api/src/main/scala/code/bankconnectors/LocalMappedConnector.scala:1332]: The whole method body should change to:
```
   val oldAccountRoutings: List[BankAccountRouting] = BankAccountRouting.findAll(By(BankAccountRouting.BankId, bankId.value),
      By(BankAccountRouting.AccountId, accountId.value))
    
    // Add or update new routing schemes
    accountRoutings.foreach(accountRouting =>
      oldAccountRoutings.find(_.AccountRoutingScheme.get == accountRouting.scheme) match {
        case Some(v) =>
          v.AccountRoutingAddress(accountRouting.address).saveMe()
        case None =>
          BankAccountRouting.create
            .BankId(bankId.value)
            .AccountId(accountId.value)
            .AccountRoutingScheme(accountRouting.scheme)
            .AccountRoutingAddress(accountRouting.address)
            .saveMe()
      }
    )

    // Delete non-present routing schemes
    oldAccountRoutings.filterNot(accountRouting => accountRoutings.exists(_.scheme == accountRouting.AccountRoutingScheme.get))
      .foreach(_.delete_!)

    (for {
      (account, _) <- LocalMappedConnector.getBankAccountCommon(bankId, accountId, callContext)
    } yield {
      account
        .kind(accountType)
        .accountLabel(accountLabel)
        .mBranchId(branchId)
        .saveMe
    }, callContext)
```

File: obp-commons/src/main/scala/com/openbankproject/commons/dto/JsonsTransfer.scala [hunk 7]
```diff
@@ -628,8 +628,7 @@ case class OutBoundUpdateBankAccount(outboundAdapterCallContext: OutboundAdapter
                               accountType: String,
                               accountLabel: String,
                               branchId: String,
-                              accountRoutingScheme: String,
-                              accountRoutingAddress: String
+                              accountRoutings: List[AccountRouting]
```
Comment (oldbig) [inline obp-commons/src/main/scala/com/openbankproject/commons/dto/JsonsTransfer.scala:631]: When modifying `obp-commons` any type, need bump up pom version:
change 3 pom.xml files's version from `1.6.3` to `1.7.0`

File: obp-api/src/main/scala/code/api/util/NewStyle.scala [hunk 8]
```diff
@@ -216,6 +216,12 @@ object NewStyle {
       }
     }
 
+    def getBankAccountByRouting(scheme: String, address: String, callContext: Option[CallContext]) : OBPReturnType[BankAccount] = {
+      Future(Connector.connector.vend.getBankAccountByRouting(scheme: String, address : String, callContext: Option[CallContext])) map { i =>
+        unboxFullOrFail(i, callContext,s"${BankAccountNotFound.replaceAll("BANK_ID and ACCOUNT_ID. ", s"scheme and address.")} Current scheme is $scheme, current address is $address", 404 )
```
[the remark this answers is not in the document]
Comment (GuillaumeKergreis) [reply obp-api/src/main/scala/code/api/util/NewStyle.scala]: Error message added

File: obp-api/src/main/scala/code/api/util/NewStyle.scala [hunk 9]
```diff
@@ -216,6 +216,12 @@ object NewStyle {
...
+      }
+    }
+
     def getBankAccountByIban(iban : String, callContext: Option[CallContext]) : OBPReturnType[BankAccount] = {
       Connector.connector.vend.getBankAccountByIban(iban : String, callContext: Option[CallContext]) map { i =>
         (unboxFullOrFail(i._1, callContext,s"${BankAccountNotFound.replaceAll("BANK_ID and ACCOUNT_ID. ", "IBAN.")} Current IBAN is $iban", 404 ), i._2)
```
[the remark this answers is not in the document]
Comment (GuillaumeKergreis) [reply obp-api/src/main/scala/code/api/util/NewStyle.scala]: Error message added

File: obp-api/src/main/scala/code/api/v3_1_0/APIMethods310.scala [hunk 10]
```diff
@@ -4811,14 +4810,28 @@ trait APIMethods310 {
...
+            }
+            alreadyExistAccountRoutings <- Future.sequence(consentJson.account_routings.map(accountRouting =>
+              NewStyle.function.getBankAccountByRouting(accountRouting.scheme, accountRouting.address, callContext)
+                .map {
+                  // If we find an already existing account routing linked to the account, it just mean we don't want to update it
+                  case bankAccount if !(bankAccount._1.bankId == bankId && bankAccount._1.accountId == accountId) => Some(accountRouting)
+                  case _ => None
+                } fallbackTo Future.successful(None)
+            ))
+            alreadyExistingAccountRouting = alreadyExistAccountRoutings.find(_.nonEmpty).flatten
```
[the remark this answers is not in the document]
Comment (GuillaumeKergreis) [reply obp-api/src/main/scala/code/api/v3_1_0/APIMethods310.scala]: Done

File: obp-api/src/main/scala/code/api/v3_1_0/APIMethods310.scala [hunk 11]
```diff
@@ -5459,6 +5472,16 @@ trait APIMethods310 {
             _ <-  Helper.booleanToFuture(InvalidISOCurrencyCode){isValidCurrencyISOCode(createAccountJson.balance.currency)}
             currency = createAccountJson.balance.currency
             (_, callContext ) <- NewStyle.function.getBank(bankId, callContext)
+            _ <- Helper.booleanToFuture(s"$InvalidAccountRoutings Duplication detected in account routings, please specify only one value per routing scheme", 400){
+              createAccountJson.account_routings.map(_.scheme).distinct.size == createAccountJson.account_routings.size
+            }
+            alreadyExistAccountRoutings <- Future.sequence(createAccountJson.account_routings.map(accountRouting =>
+              NewStyle.function.getBankAccountByRouting(accountRouting.scheme, accountRouting.address, callContext).map(_ => Some(accountRouting)).fallbackTo(Future.successful(None))
+            ))
+            alreadyExistingAccountRouting = alreadyExistAccountRoutings.find(_.nonEmpty).flatten
```
[the remark this answers is not in the document]
Comment (GuillaumeKergreis) [reply obp-api/src/main/scala/code/api/v3_1_0/APIMethods310.scala]: Done

File: obp-api/src/main/scala/code/api/v4_0_0/APIMethods400.scala [hunk 12]
```diff
@@ -1263,6 +1263,16 @@ trait APIMethods400 {
             _ <-  Helper.booleanToFuture(InvalidISOCurrencyCode){isValidCurrencyISOCode(createAccountJson.balance.currency)}
             currency = createAccountJson.balance.currency
             (_, callContext ) <- NewStyle.function.getBank(bankId, callContext)
+            _ <- Helper.booleanToFuture(s"$InvalidAccountRoutings Duplication detected in account routings, please specify only one value per routing scheme") {
+              createAccountJson.account_routings.map(_.scheme).distinct.size == createAccountJson.account_routings.size
+            }
+            alreadyExistAccountRoutings <- Future.sequence(createAccountJson.account_routings.map(accountRouting =>
+              NewStyle.function.getBankAccountByRouting(accountRouting.scheme, accountRouting.address, callContext).map(_ => Some(accountRouting)).fallbackTo(Future.successful(None))
+              ))
+            alreadyExistingAccountRouting = alreadyExistAccountRoutings.find(_.nonEmpty).flatten
```
[the remark this answers is not in the document]
Comment (GuillaumeKergreis) [reply obp-api/src/main/scala/code/api/v4_0_0/APIMethods400.scala]: Done

File: obp-api/src/main/scala/code/bankconnectors/LocalMappedConnector.scala (hunk 4 above)
Comment (GuillaumeKergreis) [reply obp-api/src/main/scala/code/bankconnectors/LocalMappedConnector.scala]: Done

File: obp-api/src/main/scala/code/bankconnectors/LocalMappedConnector.scala (hunk 5 above)
Comment (GuillaumeKergreis) [reply obp-api/src/main/scala/code/bankconnectors/LocalMappedConnector.scala]: Done

File: obp-api/src/main/scala/code/bankconnectors/LocalMappedConnector.scala (hunk 6 above)
Comment (GuillaumeKergreis) [reply obp-api/src/main/scala/code/bankconnectors/LocalMappedConnector.scala:1321]: I took the main of your proposition, just for `_.AccountRoutingScheme.get` I replaced it by `_.accountRouting.scheme`. I think it's better and don't change anything.

**Comment by GuillaumeKergreis** [comment]
Here is the last  [Jenkins pipeline with all tests passing](https://jenkins.tesobe.com/job/Build-OBP-API-guillaume-jdk13_UNLICENSED-JDK-NOT-FOR-DEPLOYMENT/12/)

**Comment by oldbig** [comment]
Code review finish.

**Comment by GuillaumeKergreis** [comment]
[Jenkins pipeline for this last commit](https://jenkins.tesobe.com/job/Build-OBP-API-guillaume-jdk13_UNLICENSED-JDK-NOT-FOR-DEPLOYMENT/14/)

**Pull request merged**